### ISSUE TITLE
Shared cloning > Change "Credits" to "Acknowledgments": Port from OpenLab

### ISSUE THREAD

**Issue by bree-z** [cuny-academic-commons/commons-in-a-box#354]
OpenLab ticket is here: http://redmine.citytech.cuny.edu/issues/2863

This is to port changes we made to the format of cloning acknowledgements on a group profile.

**Comment by bree-z** [comment]
Hi Boone, 

Acknowledgements aren't appearing correctly in all cases. For example:

1. Clone is missing acknowledgements of two ancestors:

Source: https://openlabdev.commonsinabox.org/groups/ancient-political-thought-shared-clone/
Clone: https://openlabdev.commonsinabox.org/groups/shared-clone-of-ancient-political-thought-shared-clone/

In this case, the clone isn't showing acknowledgements, but the source group does show correct acknowledgements for its ancestor group. The clone should then have acknowledgements for two ancestors: 
a) https://openlabdev.commonsinabox.org/groups/ancient-political-thought-shared-clone/
b) https://openlabdev.commonsinabox.org/groups/ancient-political-thought/

Descendants of b) do appear correctly: https://openlabdev.commonsinabox.org/groups/type/course/?descendant-of=210 

2. Another clone missing acknowledgements:

Source: https://openlabdev.commonsinabox.org/groups/summer-2021-course/
Clone: https://openlabdev.commonsinabox.org/groups/summer-2021-course-shared-clone-1/

The source group shows the correct descendent query on the group home.

Thanks!

**Comment by boonebgorges** [comment]
I'm going to move over some more changes from the OpenLab in response to this, but can you please refrain from further testing of it until it's 100% resolved at http://redmine.citytech.cuny.edu/issues/2934 (also http://redmine.citytech.cuny.edu/issues/2945)? It's very labor-intensive to do frequent moves back and forth between the two platforms, and it'll be made somewhat easier I can take a known-working version on the OL and move it here.

**Comment by boonebgorges** [comment]
A brief update to say that each of these cases has to do not with the Acknowledgements tools (which appear to be working pretty much fine) but instead the Creators data. Most of the linked groups do not have any Creators, which means the Acknowledgements cannot be properly built. I'm unsure why the Creators data is incorrect - perhaps you can reproduce your tests and take careful note of the Creators data at each step.

**Comment by bree-z** [comment]
Thanks Boone! We're working on these tickets today, and will pay careful attention to the Creators data.

**Comment by bree-z** [comment]
Hi Boone, 

I closed http://redmine.citytech.cuny.edu/issues/2934 and http://redmine.citytech.cuny.edu/issues/2945, but just making a note here too. 

Thanks!

**Comment by boonebgorges** [comment]
Thank you - I believe that I have ported over all the relevant changes from the OpenLab, so you can feel free to begin testing here. As noted above, the specific issues you most recently mentioned are due to missing Creator data, so pay attention to this as you set up your test groups.

**Comment by bree-z** [comment]
Thanks, Boone. 

I'm only seeing acknowledgements of ancestor groups when I've cloned a group and added either additional Creators or additional text to the clone. For example: 

Source: https://openlabdev.commonsinabox.org/groups/course-with-additional-creator/
Source Creator: breefaculty, breefacultyagain
Clone: https://openlabdev.commonsinabox.org/groups/clone-with-additional-creator/
Clone Creator: testfaculty

Source: https://openlabdev.commonsinabox.org/groups/course-with-additional-creator/
Source Creator: breefaculty, breefacultyagain
Clone: https://openlabdev.commonsinabox.org/groups/clone-with-additional-text/
Clone Creator: testfaculty

However, the formatting of the acknowledgements section for both of the above doesn't appear correctly. 

The [first example](https://openlabdev.commonsinabox.org/groups/clone-with-additional-creator/) should match the following (although I realize the wording needs to be slightly different to accommodate  translation):

http://openlabdev.org/groups/clone-with-additional-creator/

The [second example](https://openlabdev.commonsinabox.org/groups/clone-with-additional-text/) should match the following:

http://openlabdev.org/groups/clone-with-additional-text/

Aside from the two examples above, I haven't been able to get acknowledgments of ancestor groups to appear on group home pages of its descendants. However, ancestor queries are displaying correctly (e.g. "This course has been cloned or re-cloned 5 times.") 

Here are a few test groups:

Source: https://openlabdev.commonsinabox.org/groups/summer-2021-course/
Source Creator: testfaculty
Clone: https://openlabdev.commonsinabox.org/groups/summer-2021-course-shared-clone-1/
Clone Creator: breefaculty

Source: https://openlabdev.commonsinabox.org/groups/course-with-multiple-creators/
Source Creator: testfaculty, breefacultyagain, Kermit the Frog
Clone: https://openlabdev.commonsinabox.org/groups/b-faculty-shared-clone-of-course-with-multiple-creators/
Clone Creator: breefaculty
Clone of Clone: https://openlabdev.commonsinabox.org/groups/alvin-lanes-clone-of-b-faculty-shared-clone-of-course-with-multiple-creators/
Clone of Clone Creator: breetestaug1

Source: https://openlabdev.commonsinabox.org/groups/course-with-additional-creator/
Source Creator: breefaculty, breefacultyagain
Clone: https://openlabdev.commonsinabox.org/groups/testfaculty-shared-clone-of-course-with-additional-creator/
Clone Creator: testfaculty

Thanks!

**Comment by boonebgorges** [comment]
In the case of https://openlabdev.commonsinabox.org/groups/clone-with-additional-text/, the Acknowledgements text begins 'It is based on...'. The first part should be 'Acknowledgements: Created by: [creator names]'. But this group has no Creators, so this part of the text is not generated. See https://openlabdev.commonsinabox.org/groups/clone-with-additional-text/admin/edit-details/

The "formatting" of https://openlabdev.commonsinabox.org/groups/clone-with-additional-creator/ appears to be merely a CSS issue and not related to the content. It should now be fixed with recent changes.

In the case of https://openlabdev.commonsinabox.org/groups/summer-2021-course/admin/edit-details/ and https://openlabdev.commonsinabox.org/groups/summer-2021-course-shared-clone-1/admin/edit-details/, neither group appears to have any Creators.

https://openlabdev.commonsinabox.org/groups/course-with-multiple-creators/admin/edit-details/ has multiple creators, but https://openlabdev.commonsinabox.org/groups/b-faculty-shared-clone-of-course-with-multiple-creators/admin/edit-details/ and https://openlabdev.commonsinabox.org/groups/alvin-lanes-clone-of-b-faculty-shared-clone-of-course-with-multiple-creators/admin/edit-details/ have none.

Likewise, the source group https://openlabdev.commonsinabox.org/groups/course-with-additional-creator/admin/edit-details/ has two Creators but clone https://openlabdev.commonsinabox.org/groups/testfaculty-shared-clone-of-course-with-additional-creator/admin/edit-details/ has none.

So your description of the setup is incorrect. My guess is that something is happening during the clone process that's preventing Creators from being copied over. Either that, or something is happening during the process of you changing the Creators that's causing the information to be lost. In any case, as noted above, I do not currently believe that this is a problem with Acknowledgements per se, but is instead an issue with Creators not being populated in the way that you expect. I guess I'd recommend trying again, this time being extremely attentive to each step of the clone + settings process, noting at each step whether the Creators data is correct.

**Comment by bree-z** [comment]
Sorry about that, Boone. I wasn't checking Creators carefully enough. :-/ 

It looks like there is no Creator populated after creating a group, either via cloning or from scratch. 

- I created this group, as testfaculty, but no Creator is listed in Settings after creation: https://openlabdev.commonsinabox.org/groups/course-from-scratch/
- Logged in as breefaculty, I cloned course-from-scratch to create: https://openlabdev.commonsinabox.org/groups/bfacultys-clone-of-course-from-scratch/. There is also no Creator listed for the clone, which should have been breefaculty.
- There are also no acknowledgements of the source group listed on the clone's profile, which makes sense because the source group has no Creator.
- However, logged in as testfaculty I created this group: https://openlabdev.commonsinabox.org/groups/course-from-scratch-2/ . After creating the group, there was no Creator listed. I added testfaculty as the Creator. 
- Then, logged in as breefaculty, I cloned course-from-scratch-2 to create this group: https://openlabdev.commonsinabox.org/groups/bfacultys-clone-of-course-from-scratch-2/
- The clone has no Creator listed, but the source group did have testfaculty listed as a Creator, so shouldn't the following acknowledgements appear crediting the *source* even if bfacultys-clone-of-course-from-scratch-2 doesn't have a creator listed?

> Acknowledgements: Based on the following:
> 
> 45678 by Professor Tesla

- I did one last test, logged in as testfaculty, where I created the following group: https://openlabdev.commonsinabox.org/groups/course-from-scratch-3/ and then added testfaculty as Creator.
- Then, logged in as breefaculty, I cloned course-from-scratch-3 to create: https://openlabdev.commonsinabox.org/groups/bfacultys-clone-of-course-from-scratch-3/
- No creator was listed for bfacultys-clone-of-course-from-scratch-3. And, no acknowledgements appeared crediting course-from-scratch-3 as the source
- I added breefaculty as Creator of bfacultys-clone-of-course-from-scratch-3. After doing that, the correct acknowledgements appeared as follows:

> Acknowledgements: Based on the following:
> 
> 444 by Professor Tesla

Hopefully this helps, but let me know if you need more information. Thanks!

**Comment by boonebgorges** [comment]
Thanks for the details. It appears that the issue is due to a load-order issue. Default Creators depend on default Contacts, but Creators were being saved before the Contacts, which meant that there were no Contacts to copy over. I've done some juggling that makes this work in my tests, but you should give it a try. When you are satisfied that group creation seems to be working right, you can test the Acknowledgements stuff.

**Comment by bree-z** [comment]
Thanks, Boone!

Group creation looks good to me. I tested all group types from scratch, and also via cloning, and in each case the Creator section in Settings was populated with the person who created the group. 

There are some issues with Acknowledgements:

1. Acknowledgements don't appear for projects and clubs as they do for courses, even when the source group has a Creator. E.g.:

- Source: https://openlabdev.commonsinabox.org/groups/test-students-summer-2021-project/
- Creator (double checked in Settings): teststudent
- Clone: https://openlabdev.commonsinabox.org/groups/bfacultys-clone-of-test-students-summer-2021-project/
- Creator (double checked in Settings): breefaculty
- Source: https://openlabdev.commonsinabox.org/groups/summer-2021-club/
- Creator (double checked in Settings): teststudent
- Clone: https://openlabdev.commonsinabox.org/groups/shared-clone-of-summer-2021-club/
- Creator (double checked in Settings): testfaculty

2. When I add an additional Creator to a project or club, Acknowledgements do not appear on the Group Home, e.g.:

- https://openlabdev.commonsinabox.org/groups/plants-are-good/
- Group Contact: testfaculty 
- Creators: testfaculty (also the user who created the project), teststudent (added as Creator)
- https://openlabdev.commonsinabox.org/groups/blue-club/
- Group Contact: testfaculty 
- Creators: testfaculty (also the user who created the project), breefaculty (added as Creator)

6. When I create a group and then delete the Creator of that group, no acknowledgements should appear on descendants created via shared cloning. E.g.:
- Source: https://openlabdev.commonsinabox.org/groups/course-with-a-deleted-creator/
- Clone: https://openlabdev.commonsinabox.org/groups/shared-clone-of-course-with-a-deleted-creator/

**Comment by boonebgorges** [comment]
1 and 2 are subsets of the same problem, which is that Course headers are built differently from other group types. I've ported the Acknowledgments section to the other group types.

>  When I create a group and then delete the Creator of that group, no acknowledgements should appear on descendants created via shared cloning.

For posterity, the issue is that the Creator was removed from the list of group Creators, not that the Creator's account was deleted from the network (I got confused about this for a moment, so I wanted to make a note). The issue here is that the OL had switched its logic so that info about ancestor admins was pulled from Creators instead of Contacts, but this change was not ported to CBOX-OL. I've made the necessary change.
